File: typescript/src/structures/queue.ts



class QNode<T> {
    next: QNode<T> | null = null
    value: T

    constructor(value: T) {
        this.value = value;
    }
}

export class Queue<T> {
    private _head: QNode<T> | null = null
    private _tail: QNode<T> | null = null
    private _len: number = 0


    enqueue(value: T) {
        const node = new QNode<T>(value)

        if (this._len === 0) {
            this._head = this._tail = node
            this._len++
        }

        this._tail!.next == node
        this._tail = node;
    }


    poll(): T | null {
        if (this._len === 0) return null
        const node = this._head
        if (this._len === 1) {
            this._head = this._tail = null
            this._len--
            return node!.value
        }

        this._head = this._head!.next
        return node!.value
    }

    peek(): T | null {
        if (this._len == 0) return null;
        return this._head!.value
    }
}

File: dartlang/lib/structures/queue.dart
class _Node<T> {
  _Node<T>? next;
  final T value;

  _Node(this.value);
}

class Queue<T> {
  _Node<T>? _head;
  _Node<T>? _tail;
  int _size = 0;

  bool get isEmpty => _size == 0;

  int get size => _size;

  enqueue(T value) {
    final node = _Node<T>(value);

    if (_size == 0) {
      _head = _tail = node;
      _size++;
      return;
    }

    _tail!.next = node;
    _tail = node;
    _size++;
  }

  T? poll() {
    if (_size == 0) return null;

    final node = _head;

    if (_size == 1) {
      _head = _tail = null;
      _size--;
      return node!.value;
    }

    _head = node!.next;
    _size--;
    return node.value;
  }

  T? peek() => _head?.value;
}
